fix(exporter-collector-grpc): incorrect URL format on docs after 0.20.0 update

brunoluiz opened the pull request:
## Which problem is this PR solving?

- Fixes the connection issue mentioned at https://github.com/open-telemetry/opentelemetry-js/issues/2264, where `exporter-collector-grpc` is not connecting to otel grpc connector when using `localhost:4317`. This was partially due to an undocumented URL format, changed after 0.20.0

## Short description of the changes

- Changes documentation from `localhost:4317` to `grpc://localhost:4317` for `@opentelemetry/exporter-collector-grpc` package

brunoluiz commented:
I am a bit lost with those CI errors... I suppose these issues were not due to my changes or were they?

Flarna commented:
Looking into the asserts which fail it looks very much like it is because of the changes here
```
      Uncaught AssertionError [ERR_ASSERTION]: 'grpc://localhost:4317' === 'localhost:4317'
      + expected - actual
```

brunoluiz commented:
Seems the NodeJS 16 CI test failed. Is there a way which I can re-trigger it or is it fine as it is?

obecny commented:
Can you please ensure this is working fine with this [example](https://github.com/open-telemetry/opentelemetry-js/blob/main/examples/collector-exporter-node)

You can also update the [docker compose](https://github.com/open-telemetry/opentelemetry-js/blob/main/examples/collector-exporter-node/docker/docker-compose.yaml#L5) to the latest version you were testing.

If all is fine we tend to update [readme](https://github.com/open-telemetry/opentelemetry-js/blob/main/packages/opentelemetry-exporter-collector-grpc/README.md) with information about last tested version so that it is clear which version is compatible with this as spec and collector were not always are in sync.

dyladan commented:
> Seems the NodeJS 16 CI test failed. Is there a way which I can re-trigger it or is it fine as it is?

The node 16 test is fine, but the browser test failure is probably real.

brunoluiz commented:
I will try to run re-run the tests in composer during the weekend. The browser test is probably something that I missed while searching and replacing the default URL updates.

spencerwilson commented on packages/opentelemetry-exporter-collector-grpc/README.md:
```diff
@@ -27,7 +27,7 @@ const { BasicTracerProvider, SimpleSpanProcessor } = require('@opentelemetry/tra
 const collectorOptions = {
-  // url is optional and can be omitted - default is localhost:4317
+  // url is optional and can be omitted - default is grpc://localhost:4317
   url: '<collector-hostname>:<port>',
```
Here and elsewhere, this might be warranted
```suggestion
  url: '<collector-scheme>://<collector-hostname>:<port>',
```

brunoluiz replied:
But, for this specific case, where the collector is grpc only, shouldn't we inform the use the required scheme is `grpc`? Or does it actually support other schemes?

vmarchaud replied:
then it should be `'grpc://<collector-hostname>:<port>'` right ?

brunoluiz commented:
So, just tested it based on the examples and it seems fine. I need to note though that, on Zipkin, the service is marked as unknown instead of `BASIC-SERVICE`. I need to understand how to check the browser issue.

<img width="1344" alt="Screenshot 2021-06-18 at 10 57 17" src="https://user-images.githubusercontent.com/3219330/122543803-15519400-d024-11eb-9d97-cc9da181e3fb.png">

brunoluiz commented:
> If all is fine we tend to update readme with information about last tested version so that it is clear which version is compatible with this as spec and collector were not always in sync.
I tested with the version which is in the README currently, so should be all good 😉

brunoluiz commented:
I just re-ran the browsers tests locally through `lerna run --scope @opentelemetry/instrumentation-fetch test:browser` and it seems to work. Could someone re-trigger the browser test in the CI just to see if it is fine?

brunoluiz commented:
Ok, so something is off here. With the example, it works as far as I add the `config.url=grpc://localhost:4317`. But, if left empty and leave it to go with the default config, this will error.

Checking a bit the code, I found this: https://github.com/open-telemetry/opentelemetry-js/blame/main/packages/opentelemetry-exporter-collector-grpc/src/util.ts#L110-L122

It seems it is just used to check if the schema is either HTTP or GRPC, but actually it doesn't do anything with it, as it returns `target.host` at the end. The default value I added in this PR contains `grpc://`, which obviously breaks the GRPC client later on.

Should I revert the default values and just add documentation that `grpc://` or `http://` is required on the `url`?

Another thing as well: the grpc-collector is not sending data to the Zipkin, but I guess this is what I mentioned on issue #2264.

vmarchaud commented:
> Checking a bit the code, I found this: main/packages/opentelemetry-exporter-collector-grpc/src/util.ts#L110-L122 (blame)

I find it odd that the collector grpc check for https since it only support grpc.

> Should I revert the default values and just add documentation that grpc:// or http:// is required on the url?

I think that would be the way to go if grpc itself doesn't support using just a hostname/port combination

dyladan commented:
@brunoluiz any update on the browser tests here?

brunoluiz commented:
As mentioned on https://github.com/open-telemetry/opentelemetry-js/pull/2266#issuecomment-863951585, I've reverted the changes and just kept the README changes. The utils might have some reason to exist, but changing default values causes it to crash. I suppose more investigation might be required?

dyladan commented:
In the PR title you have `incorrect default URL`. Is the default URL actually bad or is it just the documented example with the bad URL?

brunoluiz commented:
> In the PR title you have incorrect default URL. Is the default URL actually bad or is it just the documented example with the bad URL?
After discussion here, I reverted the changes on the default URL itself, as the default URLs work fine. But, if the developer specifies some other URL, it requires the `grpc` scheme to be defined.

Due to that, I just changed title and description accordingly. Let me know if it ready to merge.

vmarchaud commented:
Another PR has been opened to update the behavior to automatically add `grpc://` if not set, but i think its make sense to tell users that its required anyway in the docs
